schema.go

```diff
@@ -106,3 +113,69 @@ func findBuildByBuilder(db *mgo.Database, builder string, buildnum int) (*LogKee
 	}
 	return build, nil
 }
+
+func findAndDeleteTests(db *mgo.Database) ([]bson.ObjectId, error) {
+	now := time.Now()
+	query := bson.M{
+		"started": bson.M{"$lte": now.Add(-deletePassedTestCutoff)},
+		"failed":  false,
+	}
+	var testIDs []bson.ObjectId
+	err := db.C("tests").Find(query).Sort("started").Limit(maxTests).Distinct("_id", &testIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "error finding tests")
+	}
+	filter := bson.M{
+		"_id": bson.M{"$in": testIDs},
+	}
+	_, err = db.C("tests").RemoveAll(filter)
+	if err != nil {
+		return nil, errors.Wrap(err, "error deleting tests")
+	}
+	return testIDs, nil
+}
+
+func deleteLogsByTests(db *mgo.Database, ids []bson.ObjectId) (*mgo.ChangeInfo, error) {
+	query := bson.M{
+		"test_id": bson.M{"$in": ids},
+	}
+	return db.C("logs").RemoveAll(query)
+}
+
+func deleteBuildsWithoutTests(db *mgo.Database) error {
+	builds := []LogKeeperBuild{}
+	err := db.C("builds").Find(bson.M{}).Select(bson.M{"_id": 1}).All(&builds)
```

Comment:
You can't read all builds into memory. For this I think you need to do an aggregation to find builds that have no tests.